examples/bond/assets/bond-dapp-stateful.py

```diff
@@ -0,0 +1,255 @@
+    # - ALGO payment from buyer to the issuer. algo_tx_price is required to cover the TX2 cost.
+    # 2. B_i ASA transfer from issuer to the buyer.
+    on_buy = Seq([
+        Assert(
+            And(
+                basic_checks,
+                # verify payment
+                Gtxn[0].type_enum() == TxnType.Payment,
+                # verify buying amount,
+                Gtxn[0].amount() >= Add(Mul(Gtxn[1].asset_amount(), App.globalGet(issue_price)), Gtxn[1].fee()),
+                # verify ASA transfer
+                Gtxn[1].type_enum() == TxnType.AssetTransfer,
+                # verify current bond is being transferred
+                Gtxn[1].xfer_asset() == App.globalGet(current_bond),
+                Gtxn[1].asset_receiver() == Gtxn[0].sender()
+            )
+        ),
+        App.globalPut(total, Add(App.globalGet(total), Gtxn[1].asset_amount())),
+        Return(Int(1))
+    ])
+
+    # Create buyback address transaction
+    # Expected arguments: [Bytes("create_buyback"), buyback address]
+    create_buyback = Seq([
+        # asserts if sender is store manager
+        Assert(
+            And(
+                Txn.sender() == App.globalGet(store_manager),
+                basic_checks
+            )
+        ),
+        # set buyback address
+        App.globalPut(Bytes("buyback"), Txn.application_args[1]),
+        Return(Int(1))
+    ])
+
+    # exit transaction
+    # Expected arguments: [Bytes("exit")]
+    on_exit = Seq([
+        Assert(
+            And(
+                basic_checks,
+                # verify first transaction is asset transfer
+                Gtxn[0].type_enum() == TxnType.AssetTransfer,
+                # verify second transaction is payment
+                Gtxn[1].type_enum() == TxnType.Payment,
+                # verify sender of first transaction is buyback address
+                Gtxn[1].sender() == App.globalGet(Bytes("buyback")),
+                # verify amount of algo received by buyer
+                Gtxn[1].amount() == Minus(Mul(Gtxn[0].asset_amount(), App.globalGet(nominal_price)), Gtxn[1].fee()),
+                Gtxn[1].receiver() == Txn.sender(),
+                # verify maturity date is passed
+                Global.latest_timestamp() > App.globalGet(maturity_date)
+            )
+        ),
+        Return(Int(1))
+    ])
+
+    # Redeem transaction
+    # Expected arguments: [Bytes("redeem_coupon")]
+    on_redeem_coupon = Seq([
+        Assert(
+            And(
+                basic_checks,
+                # User sends `B_i` to `DEX_i` lsig.
+                Gtxn[0].type_enum() == TxnType.AssetTransfer,
+                # `DEX_i` sends `B_{i+1}` to the user.
+                Gtxn[1].type_enum() == TxnType.AssetTransfer,
+                Gtxn[0].asset_amount() == Gtxn[1].asset_amount(),
+                # `Dex_i` sends coupon value to user
+                Gtxn[2].type_enum() == TxnType.Payment,
+                Gtxn[2].receiver() == Txn.sender(),
```

Comment:
Txn.sender() is same Gtxn[0].sender() in this case because redeemer called this stateful contract. therfore according to current transaction address redeemer address == Gtxn[0].sender()

Reply:
here Txn.sender() and Gtxn[0].sender() is same, Gtxn[0].sender() is more self explanatory, updated to this.